### DISCUSSION TITLE
API endpoints are executed at build time

### DISCUSSION THREAD

**Discussion by abduncan** [vercel/community#3830] [Help] [answered]
### Summary

I scaffolded a new project and added a serverless function following the Serverless Function Quick Start.

When I build the project, it executes the newly created serverless function.

This is causing an issue in another project where it is trying to initialize a Pinecone and MongoDb client.

Is this expected behavior, and if so, how do I disable it?

### Steps to Reproduce

Create a new Vercel project using the Nextjs template.

Add a serverless function following the Quickstart tutorial.

Add a console.log command at the top of the function.

Run `vercel build`.

Notice the console.log output.

**Comment by amyegan** [comment, marked as answer]
Hey @abduncan. It sounds like you're using the Next.js App Router, but please correct me if I assumed incorrectly.

The default behavior in the `app` directory is to cache as much as possible. You can use the `dynamic` [route segment configuration](https://nextjs.org/docs/app/building-your-application/routing/route-handlers#segment-config-options) option to force the route to be dynamic. Then it won't attempt to cache during the build.

In my test app, I have a `force-dynamic` example route set up like this:
```js
import { NextResponse } from 'next/server';

export const dynamic = 'force-dynamic';
 
export function GET(request) {
  console.log('Running serverless example...');
  return NextResponse.json(
    {
      body: request.body,
      path: request.nextUrl.pathname,
      query: request.nextUrl.search,
      cookies: request.cookies.getAll(),
    },
    {
      status: 200,
    },
  );
}
```

**Reply by abduncan** [reply]
Thanks, @amyegan. That resolved my issue.

Question: aren't most API routes dynamic? Isn't that the purpose of an API? Otherwise, wouldn't I create a static page?

I did notice it only happens for `GET` requests, which makes sense.

Seems like it should be by default to not cache the `api` folder by either ignoring it in the `app` folder, or having it live outside the `app` folder, and then making caching the option by importing something like `export const cacheRoute = 'cache-route'`.  I could be completely wrong, but it caught me by surprise and caused builds to fail.

**Reply by eybel** [reply]
Is this related to the "revalidate"?  mine isn't working... Just wondering if we are talking about the same thing. My post is this one: [Revalidate Issue](https://github.com/orgs/vercel/discussions/4281)
